### PR TITLE
Fix track number/burst id calculation

### PR DESCRIPTION
**Pull request by scottstanie**
This PR implements several fixes to our burst id naming scheme:

1. For ascending note crossings that happen mid-track, the track number should update (even if there is only one absolute orbit given)
2. The calculation we've been doing is actually different than ESAs, and leads to a different burst number on many frames (I believe ones closer to the crossing node are more often different since we're missing the "preamble")

For (1), We can double check our work with the manifest.safe file. In a crossing, theres two relativeOrbitNumbers:
```
(mapping) [staniewi@aurora s1-reader]$ grep -r  relativeOrbitNumber tests/
tests/data/S1A_IW_SLC__1SDV_20221024T184148_20221024T184218_045587_05735F_D6E2.SAFE/manifest.safe:            <safe:relativeOrbitNumber type="start">15</safe:relativeOrbitNumber>
tests/data/S1A_IW_SLC__1SDV_20221024T184148_20221024T184218_045587_05735F_D6E2.SAFE/manifest.safe:            <safe:relativeOrbitNumber type="stop">16</safe:relativeOrbitNumber>
```

For (2), i'm using the [Sentinel-1 Level 1 Detailed Algorithm Definition](https://sentinels.copernicus.eu/documents/247904/1877131/S1-TN-MDA-52-7445_Sentinel-1+Level+1+Detailed+Algorithm+Definition_v2-4.pdf/83624863-6429-cfb8-2371-5c5ca82907b8?t=1644837167582) section 9.25

I've also added 2 test data cases, where I included some scripts to shrink down the real data.

One note: the current test was one where the burst ids we were making were incorrect.
before: `expected_burst_id = f't071_{151200 + i}_iw3'` 

I have added a small sample of the ESA burst database with their geometries. I added a test that compares the overlap to make sure our id matches theirs.

### CODE REVIEW

**Comment by scottstanie** [comment]
Sorry if anyone started to look at this. My first pushes did fix the data that I checked (including the test data we had), but after doing an analysis of a full 12-day cycle, there were multiple other corner cases. I appear to have gotten a version that has no mismatches between ESA and all the geometries overlap with theirs.

However, with each new problem I found, I added a skeleton of that .SAFE folder with the annotations. As a result, I've got 7 now, so the `tests/data` folder is somewhat large. If people would like, I can try to put the data separately and fetch it with something like pooch: https://github.com/fatiando/pooch . 
If I did, I could also look at loading the `data/aux_cal` folder too

File: src/s1reader/s1_reader.py [hunk 1]
```diff
@@ -860,9 +850,106 @@ def _burst_from_safe_dir(safe_dir_path: str, id_str: str, orbit_path: str, flag_
     else:
         msg = f'measurement directory NOT found in {safe_dir_path}'
         msg += ', continue with metadata only.'
-        print(msg)
+        # print(msg)
```
Comment (vbrancat) [inline src/s1reader/s1_reader.py:853]: Is it intentional to have a comment here? Don't you want to just print the message for debugging purposes?

File: src/s1reader/s1_reader.py [hunk 2]
```diff
@@ -860,9 +850,106 @@ def _burst_from_safe_dir(safe_dir_path: str, id_str: str, orbit_path: str, flag_
...
+    sensing_time : datetime
+        Sensing time of the first input line of this burst [UTC]
+        The XML tag is sensingTime in the annotation file.
+    ascending_node_dt : datetime
+        Time of the ascending node prior to the start of the scene.
+    start_track : int
+        Relative orbit number at the start of the acquisition, from 1-175.
+    end_track : int
+        Relative orbit number at the end of the acquisition.
+    subswath_name : str, {'IW1', 'IW2', 'IW3'}
```
Comment (vbrancat) [inline src/s1reader/s1_reader.py:884]: Is this case-sensitive?

File: src/s1reader/s1_reader.py [hunk 3]
```diff
@@ -860,9 +850,106 @@ def _burst_from_safe_dir(safe_dir_path: str, id_str: str, orbit_path: str, flag_
...
+    The `start_track` and `end_track` parameters are used to determine if the
+    scene crosses the equator. They are the same if the frame does not cross
+    the equator.
+
+    References
+    ----------
+    ESA Sentinel-1 Level 1 Detailed Algorithm Definition
+    https://sentinels.copernicus.eu/documents/247904/1877131/S1-TN-MDA-52-7445_Sentinel-1+Level+1+Detailed+Algorithm+Definition_v2-4.pdf/83624863-6429-cfb8-2371-5c5ca82907b8
+    """
+    # Constants in Table 9-7
```
Comment (vbrancat) [inline src/s1reader/s1_reader.py:905]: Wondering again whether we should have a separate file in the repository where to insert all the constants

File: tests/data/README.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
+- `make_empty_safe.sf` converts a full SDV SAFE folder into a 2.1 Mb folder of 1 vv-polarization annotation files.
+- `make_empty_img.py` Converts a measurement/ .tiff file into an all-zeros file < 50kb using SPARSE_OK=TRUE.
```
Comment (vbrancat) [inline tests/data/README.md:2]: Not sure if it is intentional, but I do see that this PR is trying to upload a lot of files into the repository. @LiangJYu can correct me if I am wrong, but I thought we decided to upload all the test data on `zenodo.org` in order to avoid increasing the size of the s1-reader repo. If this is the case, I would recommend to submit the unit test in a separate PR as this would require some more work (i.e. data uploading, set the CI) and we desperately need this PR for our CSLC-S1 Beta delivery

**Review by vbrancat — COMMENTED** [review]
@scottstanie thanks a lot for this PR. This is so crucial to make the CSLC-S1 workflow to properly work with the TrackFrame database.

My main concern with this PR is related to the unit test. It seems that you are trying to upload a bunch of small file in the repository to make the unit test to run. If I am not mistaken, we decided to avoid going down this route. Our best option would be to store the data on `zenodo`, download them on the CI, run the unit tests. @LiangJYu and @rtburns-jpl might help you along the way.

If the above is correct, I recommend:
1. Submit the unit test in a separate PR. You might need some time to set up the repository on `zenodo` and having the test running on the CI
2. Include a README file together with the data downloaded for the unit test. This should describe the source of the data, the region covered, what files they contain and why they have been selected for the unit test

File: src/s1reader/s1_burst_slc.py [hunk 5]
```diff
@@ -652,8 +652,9 @@ def eap_compensation_lut(self):
                             f' IPF version = {self.ipf_version}')
 
         return self.burst_eap.compute_eap_compensation_lut(self.width)
-    def bbox(self):
```
[the remark this answers is not in the document]
Comment (scottstanie) [reply src/s1reader/s1_burst_slc.py:655]: I removed this once I realized why the `border` was a list of Polygon's, when it seemed like it was always a list of one single Polygon: https://github.com/opera-adt/burst_db/issues/1

the anti-meridian crossing bursts (international date line) have two polygons, since it's conventional to split the latlon that way, so this function is giving an incorrect bbox for those ones. I can leave it in and submit an issue to correct it in the future if you'd like.

[the remark this answers is not in the document]
Comment (scottstanie) [reply src/s1reader/s1_burst_slc.py:655]: fixing it might be as simple as 
```
    return shapely.geometry.MultiPolygon(b.border).bounds
```
as long as we don't care that it returns longitudes greater than 180:
```python
In [5]: b = s1reader.load_bursts('S1A_IW_SLC__1SDV_20220110T181913_20220110T181938_041402_04EC40_E2D9.zip', None, 1)[0]
In [6]: b.border
Out[6]: [<shapely.geometry.polygon.Polygon object at 0x7fadef947670>, <shapely.geometry.polygon.Polygon object at 0x7fadef944d00>]

In [8]: b.border[0].bounds
Out[8]: (180.0, 52.19675210021637, 180.6997402722514, 52.44622930532407)

In [11]: import shapely.geometry
In [12]: shapely.geometry.MultiPolygon(b.border).bounds
Out[12]: (179.3998234765543, 52.19675210021637, 180.6997402722514, 52.50961732430616)

```

File: src/s1reader/s1_reader.py (hunk 2 above)
Comment (scottstanie) [reply src/s1reader/s1_reader.py:884]: nope, and I'll add a comment saying so

File: src/s1reader/s1_reader.py (hunk 1 above)
Comment (scottstanie) [reply src/s1reader/s1_reader.py:853]: ah sorry, I had turned this off while downloading the cycle of metadata so it didn't print the message 3 * 10 * (number of frames in a cycle) times. I'll turn it back on, but it would be nice to have it be a `.debug` logging message once we switch in python logging or do #25

File: src/s1reader/s1_reader.py [hunk 6]
```diff
@@ -270,21 +267,33 @@ def get_burst_centers_and_boundaries(tree):
...
+
+def _get_manifest_pattern(tree: ET, keys: list):
+    '''Get the search path to extract data from the ET of manifest.safe'''
+    # Get the namespace from the root element to avoid full urls
+    try:
+        nsmap = tree.nsmap
+    except AttributeError:
+        nsmap = tree.getroot().nsmap
+    # path to xmlData in manifest
+    xml_meta_path = 'metadataSection/metadataObject/metadataWrap/xmlData'
```
[the remark this answers is not in the document]
Comment (scottstanie) [reply src/s1reader/s1_reader.py:292]: I can do that, but do we think we'll use it elsewhere in the code? My thought was that if it's used just once in a function, moving it away from that function would make it harder to debug if something's wrong.

[the remark this answers is not in the document]
Comment (scottstanie) [reply src/s1reader/s1_reader.py:292]: Actually if we did that, do you think we should move the XML path strings that are at the top of the `burst_from_xml` too into something like a "schema" file? I'm talking about the ones like 
```
tree.find('generalAnnotation/productInformation')
...
tree.find('imageAnnotation/imageInformation')
...
tree.find('generalAnnotation/downlinkInformationList/downlinkInformation')
...
```

File: src/s1reader/s1_reader.py (hunk 3 above)
Comment (scottstanie) [reply src/s1reader/s1_reader.py:931]: Same comment here- I can move these to a constants file if you'd like, but i'm unsure if the constants would be easier to interpret if they're moved away from function that uses them. If we're planning on making a big `constants.py` then I can start with these

File: MANIFEST.in [hunk 7]
```diff
@@ -1 +1 @@
-include src/s1reader/data/sentinel1_track_burst_id.txt
+recursive-include src/s1reader/data/ *
```
Comment (scottstanie) [inline MANIFEST.in:1]: As a note: this will fix #85 by including all the files in the data folder

**Comment by scottstanie** [comment]
alright @vbrancat now this PR is much smaller haha. i've separated out the unit tests/new test file generation into https://github.com/opera-adt/s1-reader/pull/84 and i'll talk with @LiangJYu about his usage of Zenodo-pulled files.

File: tests/create_esa_db_sample.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
+import pandas as pd
```
[the remark this answers is not in the document]
Comment (scottstanie) [reply tests/create_esa_db_sample.py]: ok i'll keep that in mind for the other PR that I moved these test scripts to 👍

File: src/s1reader/s1_reader.py [hunk 9]
```diff
@@ -866,3 +882,100 @@ def _burst_from_safe_dir(safe_dir_path: str, id_str: str, orbit_path: str, flag_
...
+    ]
+    offset = iw1_start_offsets[swath_num - 1]
+    start_iw1 = sensing_time + datetime.timedelta(seconds=offset)
+
+    start_iw1_to_mid_iw2 = burst_times[0] + burst_times[1] / 2
+    mid_iw2 = start_iw1 + datetime.timedelta(seconds=start_iw1_to_mid_iw2)
+
+    has_anx_crossing = (end_track == start_track + 1) or (
+        end_track == 1 and start_track == 175
+    )
```
[the remark this answers is not in the document]
Comment (scottstanie) [reply src/s1reader/s1_reader.py:958]: haha I think i waffled back and forth between whether the modulo was clearer or the other one. i'll change to the modulo version

[the remark this answers is not in the document]
Comment (scottstanie) [reply src/s1reader/s1_reader.py:958]: (idk why i changed it back from this commit https://github.com/opera-adt/s1-reader/blob/dba751c6247481ebc5531ffaf11566bd03a25d34/src/s1reader/s1_reader.py#L890 )

File: src/s1reader/s1_reader.py [hunk 10]
```diff
@@ -268,23 +265,59 @@ def get_burst_centers_and_boundaries(tree):
...
+    tree : xml.etree.ElementTree
+        ElementTree containing the parsed 'manifest.safe' file.
+    keys : list
+        List of keys to search for in the manifest file.
+
+    Returns
+    -------
+    str
+        Search path to extract from the ET of the manifest.safe XML.
+
```
[the remark this answers is not in the document]
Comment (scottstanie) [reply src/s1reader/s1_reader.py:308]: good idea, i added a comment and a link to the lxml document where it came from

File: src/s1reader/s1_reader.py [hunk 11]
```diff
@@ -866,3 +885,99 @@ def _burst_from_safe_dir(safe_dir_path: str, id_str: str, orbit_path: str, flag_
...
     return bursts
+
+
+def get_burst_id(
+    sensing_time: datetime.datetime,
+    ascending_node_dt: datetime.datetime,
+    start_track: int,
+    end_track: int,
+    subswath_name: str,
+) -> int:
```
Comment (LiangJYu) [inline src/s1reader/s1_reader.py:896]:
```suggestion
) -> str:
```

File: src/s1reader/s1_reader.py [hunk 12]
```diff
@@ -866,3 +885,99 @@ def _burst_from_safe_dir(safe_dir_path: str, id_str: str, orbit_path: str, flag_
...
+        Relative orbit number at the end of the acquisition.
+    subswath_name : str, {'IW1', 'IW2', 'IW3'}
+        Name of the subswath of the burst (not case sensitive).
+
+    Returns
+    -------
+    relative_orbit : int
+        Relative orbit number (track number) at the current burst.
+    burst_id : int
+        The burst ID matching ESA's relative numbering scheme.
```
Comment (LiangJYu) [inline src/s1reader/s1_reader.py:921]:
```suggestion
    burst_id : str
        The burst ID string made up of track numer + ESA's relative numbering scheme + swath ID.
```

File: src/s1reader/s1_reader.py (hunk 3 above)
Comment (LiangJYu) [reply src/s1reader/s1_reader.py:934]: I don't think there's enough for a dedicated constants file. From a quick glance, I didn't see anything beyond 3 constants below in the repo. 

What do you about moving `get_burst_id` into its own file? A `s1_burst_id.py` that's something like the following (some inspiration [from](https://stackoverflow.com/questions/61937520/proper-way-to-create-class-variable-in-data-class)):
```python
from typing import ClassVar
from dataclasses import dataclass

@dataclass
class S1BurstID:
   T_beam: ClassVar[float] = 2.758273  # interval of one burst [s]
    T_pre: ClassVar[float] = 2.299849  # Preamble time interval [s]
    T_orb: ClassVar[float] = 12 * 24 * 3600 / 175  # Nominal orbit period [s]
    track_number: int
    esa_burst_id: int
    subswath_name: str

    @classmethod
    def params_to_id(cls, 
        sensing_time: datetime.datetime, 
        ascending_node_dt: datetime.datetime,
        start_track: int,
        end_track: int,
        subswath_name: str):
        # do same computations
        return cls(track_numer, esa_burst_id, subswath)

    @property
    def as_str(self):
        return f"t{self.track_number:03d}_{self.esa_burst_id:06d}_{self.subswath_name.lower()}"
```

We can then change [this](https://github.com/opera-adt/s1-reader/blob/cf857643af13a811aaa7c5f4ce21d54f0f093290/src/s1reader/s1_burst_slc.py#L147) from `str` to `S1BurstID`

By importing this module, all the constants as accessible a class properties.

File: src/s1reader/s1_reader.py [hunk 13]
```diff
@@ -590,31 +616,28 @@ def burst_from_xml(annotation_path: str, orbit_path: str, tiff_path: str,
         osv_list = []
 
     # load individual burst
+    d_seconds = 0.5 * (n_lines - 1) * azimuth_time_interval
```
Comment (LiangJYu) [inline src/s1reader/s1_reader.py:619]:
```suggestion
    half_burst_in_seconds = 0.5 * (n_lines - 1) * azimuth_time_interval
```
nit - something more descriptive? It was not immediately obvious what `d_seconds` represented.

File: src/s1reader/s1_reader.py [hunk 14]
```diff
@@ -761,10 +781,10 @@ def load_bursts(path: str, orbit_path: str, swath_num: int, pol: str = 'vv',
     if burst_ids:
         bursts = [b for b in bursts if b.burst_id in burst_ids]
 
-        burst_ids_found = set([b.burst_id for b in bursts])
+        burst_ids_found = set([str(b.burst_id) for b in bursts])
```
Comment (LiangJYu) [inline src/s1reader/s1_reader.py:784]:
```suggestion
        # convert S1BurstId to string for consistent object comparison later
        burst_ids_found = set([str(b.burst_id) for b in bursts])
```
nit - for clarity

File: src/s1reader/s1_reader.py [hunk 15]
```diff
@@ -761,10 +781,10 @@ def load_bursts(path: str, orbit_path: str, swath_num: int, pol: str = 'vv',
...
 
         warnings.simplefilter("always")
-        set_burst_ids = set(burst_ids)
+        set_burst_ids = set([str(b) for b in burst_ids])
```
Comment (LiangJYu) [inline src/s1reader/s1_reader.py:787]:
```suggestion
        # burst IDs as strings for consistent object comparison
        set_burst_ids = set([str(b) for b in burst_ids])
```
nit - clarity

**Comment by scottstanie** [comment]
@LiangJYu the new module is up, where you make the `S1BurstId` with the `.from_burst_params` classmethod

we'll have to change this bit in COMPASS

https://github.com/opera-adt/COMPASS/blob/bb2630675878d8a227e1c0eeb0ea1dd0266c378e/src/compass/utils/runconfig.py#L395-L400

to `'_'.join(str(b.burst_id))` to join the burst id to other strings

**Review by LiangJYu — APPROVED** [review]
LGTM. Really nice changes/additions to s1-reader!

File: src/s1reader/s1_reader.py (hunk 13 above)
Comment (scottstanie) [reply src/s1reader/s1_reader.py:619]: sure i'll change that, i moved it from the currently named variable https://github.com/opera-adt/s1-reader/blob/main/src/s1reader/s1_reader.py#L612

**Comment by scottstanie** [comment]
also, perhaps I should open an issue in RTC for @gshiroma for the same small fix once this is merged in: https://github.com/opera-adt/RTC/blob/93c3118ea4cb91058c4afc0a9e410444b038a57f/src/rtc/runconfig.py#L428 which will just have to coerce to `str()` to combine using `join()`

**Comment by scottstanie** [comment]
although I actually see one other change to the RTC repo: https://github.com/opera-adt/RTC/blob/cf04dc7ff68f559d3fa56605321c312a144b3e7b/src/rtc/h5_prep.py#L144

@LiangJYu do you think
1. we should add a method to `S1BurstId` that matches the string.split method
```python
def split(sep=None, maxsplit=-1):
    return str(self).split(sep, maxsplit)
```
2. we should have people explicitly do `str(burst_id_obj)` to use the string methods on the string repr of S1BurstId?

**Comment by LiangJYu** [comment]
> although I actually see one other change to the RTC repo: https://github.com/opera-adt/RTC/blob/cf04dc7ff68f559d3fa56605321c312a144b3e7b/src/rtc/h5_prep.py#L144
> 
> @LiangJYu do you think
> 
>     1. we should add a method to `S1BurstId` that matches the string.split method
> 
> 
> ```python
> def split(sep=None, maxsplit=-1):
>     return str(self).split(sep, maxsplit)
> ```
> 
>     2. we should have people explicitly do `str(burst_id_obj)` to use the string methods on the string repr of S1BurstId?


With regards to the RTC code, I think it would be clearer if it `track_number` was retrieved via `burst_in.burst_id.track_number` instead of `string.split()`. This train of thought sort of leads to me thinking we should not add a split method.

What other string methods were you think about besides `split`? Comparisons for individual attributes?

**Comment by scottstanie** [comment]
i didn't have any others; I was just checking thoughts on trying to not break other code now that are expected it to be a string, vs. having everyone switch to use the new object

**Comment by gshiroma** [comment]
> i didn't have any others; I was just checking thoughts on trying to not break other code now that are expected it to be a string, vs. having everyone switch to use the new object

Thank you, @scottstanie and @LiangJYu ! No problem if you guys decide to go ahead and make `burst_id` an object (other than `str`). We'll update our code accordingly. The fact that you @scottstanie identified the points of change and @LiangJYu suggested a solution makes it even easier to fix. Thanks again!

File: src/s1reader/s1_burst_slc.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (vbrancat) [reply src/s1reader/s1_burst_slc.py:689]: Let's submit an issue separately. We do not want to have longitudes greater than 180 (if this is the only fix)

File: src/s1reader/s1_reader.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vbrancat) [reply src/s1reader/s1_reader.py:321]: I guess you are right. If it is only used once, it is not worth to go through such pain :D

**Review by vbrancat — APPROVED** [review]
@scottstanie Thanks for having addressed my comments. I have no comment left at this stage :). Nice work.
